### PR TITLE
feat: wl-456 drawer

### PR DESCRIPTION
**Pull request by dev-tronghieu**

### CODE REVIEW

File: packages/app/features/DashboardLayout/Navigator.tsx [hunk 1]
```diff
@@ -88,6 +88,7 @@ const styles = StyleSheet.create({
 	container: {
 		paddingVertical: 12,
 		backgroundColor: '#131C24',
+		height: '100%',
```
Comment (FME849) [inline packages/app/features/DashboardLayout/Navigator.tsx:91]: is it possible to use `flex: 1` instead?

File: apps/mobile/src/components/DrawerNavigation/Sidebar.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+export const sidebarWidth = 64;
+
+const IS_HARDCODED = true;
+
+export const Sidebar: FC<DrawerContentComponentProps> = ({ navigation }) => {
+	const { profile, activeWidgetId } = useSnapshot(appState);
+	const widgets = useWidgets();
+
+	const handleExtensionPress = (item: WidgetDocument) => {
+		const id = item._id === '' ? 'Explore' : item._id;
```
Comment (FME849) [inline apps/mobile/src/components/DrawerNavigation/Sidebar.tsx:24]: can this be reduced to `const id = item._id || 'Explore'`?

File: apps/mobile/src/components/DrawerNavigation/Sidebar.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		navigation.navigate(id);
+	};
+
+	const handleRemoveLayout = async (layout: WidgetDocument) => {
+		await modules.storage.put(layout);
+		await navigation.navigate('Explore');
+	};
+
+	const getRouteActive = (item: WidgetDocument) => {
+		return activeWidgetId === (item._id === '' ? 'Explore' : item._id);
```
Comment (FME849) [inline apps/mobile/src/components/DrawerNavigation/Sidebar.tsx:35]: the same as above

File: apps/mobile/src/components/DrawerNavigation/Sidebar.tsx [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	const widgets = useWidgets();
+
+	const handleExtensionPress = (item: WidgetDocument) => {
+		const id = item._id === '' ? 'Explore' : item._id;
+		appActions.setActiveWidgetRoute(id);
+		navigation.navigate(id);
+	};
+
+	const handleRemoveLayout = async (layout: WidgetDocument) => {
+		await modules.storage.put(layout);
```
Comment (FME849) [inline apps/mobile/src/components/DrawerNavigation/Sidebar.tsx:30]: I think we should avoid using storage methods directly, try to unify actions related to similar entity (i.e tokens, widgets) under one object, reference `appActions`, `tokenActions` etc.

File: apps/mobile/src/components/TabNavivation/TabBar.tsx [hunk 5]
```diff
@@ -1,14 +1,37 @@
...
-export const BottomNavigationTabBar: FC<BottomTabBarProps> = ({
+
+const getIconImage = (routeName: string): ImageSourcePropType | undefined => {
+	switch (routeName) {
+		case 'Profile':
+			return { uri: appState.profile.profileImage };
+		case 'Explore':
+			return require('assets/img/explore.png');
+		case 'OurProject':
+			return require('assets/img/icon-white.png');
```
Comment (FME849) [inline apps/mobile/src/components/TabNavivation/TabBar.tsx:20]: put these require into `apps/mobile/src/utils/config.ts`

File: apps/mobile/src/stacks/index.tsx [hunk 6]
```diff
@@ -25,23 +33,44 @@ export const AppStack: FC<StackScreenProps<RootParamList>> = () => {
...
-				</Stack.Navigator>
+				{isLoggedIn ? (
+					<Drawer.Navigator
+						drawerContent={Sidebar}
+						screenOptions={{
+							headerShown: false,
+							drawerStyle: {
+								width: sidebarWidth,
+								backgroundColor: 'transparent',
+								marginBottom: tabBarHeight,
```
Comment (tanlethanh) [inline apps/mobile/src/stacks/index.tsx:44]: use StyleSheet to create style

File: apps/mobile/src/components/DrawerNavigation/Sidebar.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import type { FC } from 'react';
+import { View } from 'react-native';
+import type { DrawerContentComponentProps } from '@react-navigation/drawer';
+import DashboardNavigator from '@walless/app/features/DashboardLayout/Navigator';
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/DrawerNavigation/Sidebar.tsx:4]: don't use this, export `DashboardNavigator` in `@walless/app`, and then:
import { DashboardNavigator } from '@walless/app'

File: apps/mobile/src/components/DrawerNavigation/Sidebar.tsx [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+import {
+	useSafeAreaInsets,
+	useSnapshot,
+	useWidgets,
+} from '@walless/app/utils/hooks';
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/DrawerNavigation/Sidebar.tsx:9]: don't use this kind of path, import from `@walless/app` instead - if anything not available yet at package level, go export it from within the package.

File: apps/mobile/src/components/DrawerNavigation/Sidebar.tsx [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+import { appState, mockWidgets, widgetActions } from '@walless/engine';
+import type { WidgetDocument } from '@walless/store';
+import { appActions } from 'state/app';
+
+export const sidebarWidth = 64;
+
+const IS_HARDCODED = true;
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/DrawerNavigation/Sidebar.tsx:16]: we may not need this flag, just directly use `mockWidgets` if we did not have real one -- for local debug go change the `widgets` prop directly.

File: apps/mobile/src/components/DrawerNavigation/Sidebar.tsx [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+		appActions.setActiveWidgetRoute(id);
+		navigation.navigate(id);
+	};
+
+	const handleRemoveWidget = async (widget: WidgetDocument) => {
+		await widgetActions.removeWidget(widget);
+		await navigation.navigate('Explore');
+	};
+
+	const getRouteActive = (item: WidgetDocument) => {
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/DrawerNavigation/Sidebar.tsx:33]: should have name: `getActiveRoute`

File: apps/mobile/src/components/DrawerNavigation/Sidebar.tsx [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+		return activeWidgetId === (item._id || 'Explore');
+	};
+
+	const { top } = useSafeAreaInsets();
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/DrawerNavigation/Sidebar.tsx:37]: hooks should be on top, together in a single block -- move this line above next to `useWidgets()`

File: apps/mobile/src/components/DrawerNavigation/Sidebar.tsx [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+	return (
+		<View style={{ marginTop: top }}>
+			<DashboardNavigator
+				profile={profile}
+				widgets={IS_HARDCODED ? mockWidgets : widgets}
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/DrawerNavigation/Sidebar.tsx:43]: why we need IS_HARDCODED flag?

File: apps/mobile/src/components/TabNavivation/Item.tsx [hunk 13]
```diff
@@ -35,4 +46,13 @@ const styles = StyleSheet.create({
 	tabTitle: {
 		textAlign: 'right',
 	},
+	icon: {
+		borderRadius: 1000,
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/TabNavivation/Item.tsx:50]: don't over use radius value like this, it won't work on certain case, use `height / 2` instead.

File: apps/mobile/src/components/TabNavivation/TabBar.tsx [hunk 14]
```diff
@@ -1,14 +1,37 @@
 import type { FC } from 'react';
-import type { ViewStyle } from 'react-native';
+import type { ImageSourcePropType, ViewStyle } from 'react-native';
 import { StyleSheet, View } from 'react-native';
 import type { BottomTabBarProps } from '@react-navigation/bottom-tabs';
+import { useDrawerStatus } from '@react-navigation/drawer';
+import type { DrawerNavigationHelpers } from '@react-navigation/drawer/lib/typescript/src/types';
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/TabNavivation/TabBar.tsx:6]: don't import from this kind of path

File: apps/mobile/src/components/TabNavivation/TabBar.tsx [hunk 15]
```diff
@@ -1,14 +1,37 @@
...
+const getIconImage = (routeName: string): ImageSourcePropType | undefined => {
+	switch (routeName) {
+		case 'Profile':
+			return { uri: appState.profile.profileImage };
+		case 'Explore':
+			return nativeAsset.tabBar.explore;
+		case 'OurProject':
+			return nativeAsset.tabBar.walless;
+		default:
+			break;
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/TabNavivation/TabBar.tsx:22]: I don't think we have the same convention on Extension yet -- will this work together with Extension implementation?

File: apps/mobile/src/components/TabNavivation/TabBar.tsx [hunk 16]
```diff
@@ -1,14 +1,37 @@
 import type { FC } from 'react';
-import type { ViewStyle } from 'react-native';
+import type { ImageSourcePropType, ViewStyle } from 'react-native';
 import { StyleSheet, View } from 'react-native';
 import type { BottomTabBarProps } from '@react-navigation/bottom-tabs';
+import { useDrawerStatus } from '@react-navigation/drawer';
+import type { DrawerNavigationHelpers } from '@react-navigation/drawer/lib/typescript/src/types';
 import type { Route } from '@react-navigation/native';
+import { appActions, appState } from 'state/app';
+import { nativeAsset } from 'utils/config';
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/TabNavivation/TabBar.tsx:9]: don't use nativeAsset directly like this, import `assets` from `@walless/ioc` instead.

File: apps/mobile/src/components/TabNavivation/TabBar.tsx [hunk 17]
```diff
@@ -22,21 +45,36 @@ export const BottomNavigationTabBar: FC<BottomTabBarProps> = ({
...
+	const drawerStatus = useDrawerStatus();
+
 	return (
 		<View style={[styles.container, containerStyle]}>
+			<NavigationItem
+				route={{
+					key: 'OurProject',
+					name: 'OurProject',
+					params: {} as never,
+				}}
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/TabNavivation/TabBar.tsx:62]: did not seems that you've completed implement this piece - should not be static I guess

File: apps/mobile/src/stacks/index.tsx [hunk 18]
```diff
@@ -25,23 +33,40 @@ export const AppStack: FC<StackScreenProps<RootParamList>> = () => {
...
+							component={SplashScreen}
+							options={screenOptions.fade}
+						/>
+						<Stack.Screen
+							name="Authentication"
+							component={AuthenticationStack}
+							options={screenOptions.fade}
+						/>
+					</Stack.Navigator>
+				)}
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/stacks/index.tsx:69]: DrawerStack did not override or conditionally rendered like this. Please keep the old navigation structure, and register Drawer inside `DashboardStack` -- only users who passed sign-in see `DashboardStack` together with the DrawerStack.

File: apps/mobile/src/stacks/index.tsx [hunk 19]
```diff
@@ -13,10 +17,14 @@ import AuthenticationStack from './Authentication';
...
 const Stack = createStackNavigator<RootParamList>();
+const Drawer = createDrawerNavigator();
 
-export const AppStack: FC<StackScreenProps<RootParamList>> = () => {
+export const AppStack = () => {
 	const modalContainerRef = useRef<View>(null);
 	const theme = useSnapshot(themeState);
+	const { profile } = useSnapshot(appState);
+
+	const isLoggedIn = !!profile.id;
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/stacks/index.tsx:27]: we don't need this!

File: apps/mobile/src/state/app/index.ts [hunk 20]
```diff
@@ -10,6 +10,9 @@ export const appActions = {
 	setLoading: (flag: boolean): void => {
 		appState.loading = flag;
 	},
+	setActiveWidgetRoute: (id: string): void => {
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/state/app/index.ts:13]: should have name: `setActiveWidget`

File: apps/mobile/src/utils/config.ts [hunk 21]
```diff
@@ -16,6 +16,14 @@ export const nativeAsset: UniversalAsset = {
...
 			cardBackground: require('assets/img/widget/tezos-card-bg.png'),
 		},
+		aptos: {
+			cardIcon: require('assets/img/widget/aptos-icon.png'),
+			cardBackground: require('assets/img/widget/sky-card-bg.png'),
+		},
+	},
+	tabBar: {
+		explore: require('assets/img/explore.png'),
+		walless: require('assets/img/icon-white.png'),
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/utils/config.ts:26]: I know the reason, but this Aptos stuff shouldn't be part of this PR.

File: apps/mobile/src/utils/navigation.ts [hunk 22]
```diff
@@ -55,15 +58,15 @@ export const linking: LinkingOptions<RootParamList> = {
...
-					Home: {
-						path: '/',
-						screens: {
-							Widget: '/widget/:id',
-						},
-					},
 					Explore: '/explore',
 					Profile: '/profile',
-					Setting: '/setting',
+				},
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/utils/navigation.ts:63]: rethink about the navigation structure!

File: apps/web/src/screens/Dashboard/Aptos.tsx
Comment (fluctlight-kayaba) [inline apps/web/src/screens/Dashboard/Aptos.tsx:1]: this change should not be here

File: packages/app/features/DashboardLayout/Navigator.tsx [hunk 23]
```diff
@@ -40,10 +41,26 @@ export const DashboardNavigator: FC<Props> = ({
 		} as never,
 	};
 	const isExplorerActive = getIsExtensionActive?.(exploreItem as never);
+	const { isMobile } = runtime;
```
Comment (fluctlight-kayaba) [inline packages/app/features/DashboardLayout/Navigator.tsx:44]: no platform-conditional here, please revise this with me @dev-tronghieu

File: packages/app/features/DashboardLayout/NavigatorOrb.tsx [hunk 24]
```diff
@@ -49,7 +50,9 @@ export const NavigatorOrb: FC<Props> = ({
 	const iconColor = getIconColor(isActive, item.storeMeta);
 	const iconSize = item.storeMeta?.iconSize || 20;
 	const iconUri = item.storeMeta?.iconUri;
-	const iconSource = { uri: iconUri };
+	const iconSource = runtime.isMobile
```
Comment (fluctlight-kayaba) [inline packages/app/features/DashboardLayout/NavigatorOrb.tsx:53]: shouldn't be conditional this way

File: packages/app/features/Widget/shared.ts [hunk 25]
```diff
@@ -19,6 +19,7 @@ export const widgetMap: Record<string, WidgetComponent> = {
 	sui: BuiltInNetwork,
 	tezos: BuiltInNetwork,
 	solana: BuiltInNetwork,
+	aptos: BuiltInNetwork,
```
Comment (fluctlight-kayaba) [inline packages/app/features/Widget/shared.ts:22]: not belongs to this PR

File: packages/engine/crawlers/aptos/index.ts
Comment (fluctlight-kayaba) [inline packages/engine/crawlers/aptos/index.ts:1]: not part of this PR

File: packages/engine/index.ts
Comment (fluctlight-kayaba) [inline packages/engine/index.ts:1]: not part of this PR

File: apps/mobile/src/components/DrawerNavigation/Sidebar.tsx (hunk 12 above)
Comment (dev-tronghieu) [reply apps/mobile/src/components/DrawerNavigation/Sidebar.tsx:43]: we need to handle the selected widgets of the user programmatically, for now in mobile we have not implemented it yet hence the `IS_HARDCODED` is used to remind us to remove it later

File: apps/mobile/src/components/DrawerNavigation/Sidebar.tsx (hunk 9 above)
Comment (dev-tronghieu) [reply apps/mobile/src/components/DrawerNavigation/Sidebar.tsx:16]: I've replied in the below conversation

File: apps/mobile/src/components/TabNavivation/TabBar.tsx (hunk 14 above)
Comment (dev-tronghieu) [reply apps/mobile/src/components/TabNavivation/TabBar.tsx:6]: the `DrawerNavigationHelpers` is exposed in this lib path, might I re-export it in another the file (such as `@walless/abc`) and import from that file for a prettier import (`import type { DrawerNavigationHelpers } from '@walless/abc'`)

File: apps/mobile/src/components/TabNavivation/TabBar.tsx (hunk 15 above)
Comment (dev-tronghieu) [reply apps/mobile/src/components/TabNavivation/TabBar.tsx:22]: I don't think this one is related to the extension, it is the Bottom Tab icons which only used on mobile, I guess?

File: apps/mobile/src/components/TabNavivation/TabBar.tsx (hunk 17 above)
Comment (dev-tronghieu) [reply apps/mobile/src/components/TabNavivation/TabBar.tsx:62]: Following the design, we have a Walless icon used to toggle the drawer and it is placed in the bottom tabs such as the other tabs, but it does not has a tab behavior. That's why I let it empty, do you have any suggestions?

File: apps/mobile/src/components/DrawerNavigation/Sidebar.tsx (hunk 9 above)
Comment (fluctlight-kayaba) [reply apps/mobile/src/components/DrawerNavigation/Sidebar.tsx:11]: as discussed offline, still need to remove `IS_HARDCODED` flag

File: apps/mobile/src/components/DrawerNavigation/index.tsx [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+			initialRouteName="Explore"
+		>
+			<Drawer.Screen name="Explore" component={ExploreScreen} />
+			{mockWidgets.map((widget) => (
+				<Drawer.Screen
+					key={widget._id}
+					name={widget._id}
+					component={WidgetScreen}
+				/>
+			))}
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/DrawerNavigation/index.tsx:39]: we don't want to pre-render all Widgets using Navigation system, have a look at [this reference](https://github.com/cocrafts/walless/blob/5d907b40c550f5bcf267f37d70231556f924d6ce/apps/web/src/screens/Dashboard/Embed.tsx#L13).

File: apps/mobile/src/components/TabNavivation/Item.tsx [hunk 27]
```diff
@@ -1,25 +1,36 @@
...
-			<Text style={styles.tabTitle}>{route.name}</Text>
+			{tabIcon ? (
+				<Image
+					source={tabIcon}
+					resizeMode="contain"
+					style={[styles.icon, focused && styles.iconActive]}
+				/>
+			) : (
+				<Text style={styles.tabTitle}>{route.name}</Text>
+			)}
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/TabNavivation/Item.tsx:33]: we'll force `Icon`, there would no `Text` support at all.

File: apps/mobile/src/components/TabNavivation/TabBar.tsx [hunk 28]
```diff
@@ -1,17 +1,46 @@
-import type { FC } from 'react';
-import type { ViewStyle } from 'react-native';
+import { type FC, useEffect, useState } from 'react';
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/components/TabNavivation/TabBar.tsx:1]:
```typescript
import type { FC } from 'react';
import { useEffect, useState } from 'react';
```

File: packages/app/features/DashboardLayout/Navigator.tsx (hunk 23 above)
Comment (dev-tronghieu) [reply packages/app/features/DashboardLayout/Navigator.tsx:44]: as I use the same `DashboardNavigator` component for both mobile and extension and their UIs are not identical, I have to check for the platform to manage these differences

File: packages/app/features/DashboardLayout/NavigatorOrb.tsx (hunk 24 above)
Comment (dev-tronghieu) [reply packages/app/features/DashboardLayout/NavigatorOrb.tsx:53]: please guide me to refactor this

File: apps/mobile/src/stacks/index.tsx [hunk 29]
```diff
@@ -1,20 +1,20 @@
...
 
+import { BottomTabNavigation } from '../components/TabNavivation';
+
 import AuthenticationStack from './Authentication';
-import DashboardStack from './Dashboard';
 
 const Stack = createStackNavigator<RootParamList>();
 
-export const AppStack: FC<StackScreenProps<RootParamList>> = () => {
+export const AppStack = () => {
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/stacks/index.tsx:17]: why remove this? we'll need it

File: apps/mobile/src/stacks/index.tsx [hunk 30]
```diff
@@ -38,7 +38,7 @@ export const AppStack: FC<StackScreenProps<RootParamList>> = () => {
 					/>
 					<Stack.Screen
 						name="Dashboard"
-						component={DashboardStack}
+						component={BottomTabNavigation}
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/stacks/index.tsx:41]: in this place, `DashboardStack` would be much nicer name versus `BottomTabNavigation` as we're mentioning an entire Stack/FeatureBundle instead of Navigation.

File: apps/mobile/src/utils/navigation.ts [hunk 31]
```diff
@@ -21,14 +21,9 @@ export type HomeParamList = {
 };
 
 export type DashboardParamList = {
-	Home: NavigatorScreenParams<HomeParamList>;
+	Home: undefined;
```
Comment (fluctlight-kayaba) [inline apps/mobile/src/utils/navigation.ts:24]: why remove this?

**Review by fluctlight-kayaba — APPROVED** [review]
